src/wallet/wallet.cpp

```diff
@@ -1464,9 +1466,30 @@ void CWallet::blockConnected(ChainstateRole role, const interfaces::BlockInfo& b
     // Uses chain max time and twice the grace period to adjust time for block time variability.
     if (block.chain_time_max < m_birth_time.load() - (TIMESTAMP_WINDOW * 2)) return;
 
+    // Retrieve the undo data from disk
+    // Although BlockInfo has a field for the undo data, it actually is not provided by blockConnected,
+    // so we need to pull it up from disk.
+    // Since blocks with only one tx do not have any relevant undo data, we can avoid unnecessary disk I/O
+    // by skipping this step for those blocks.
+    CBlockUndo block_undo;
+    if (block.data->vtx.size() > 0 && IsWalletFlagSet(WALLET_FLAG_SILENT_PAYMENTS)) {
+        chain().findBlock(block.hash, FoundBlock().undoData(block_undo));
+    }
+
     // Scan block
     bool wallet_updated = false;
     for (size_t index = 0; index < block.data->vtx.size(); index++) {
+        CTransactionRef tx{block.data->vtx.at(index)};
```

Comment:
Do we need the `CTransactionRef` here, as opposed to just operating on `block.data->vtx` directly?

Reply:
We only use the reference once, so it's not needed. I'll take this out as I retouch.

Reply:
I ended up leaving the reference and used it in 2 other locations. The code looks slightly cleaner this way.